model/testCenter/TestCenterService.php

```diff
@@ -0,0 +1,114 @@
+<?php
+/**
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; under version 2
+ * of the License (non-upgradable).
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * Copyright (c) 2019 (original work) Open Assessment Technologies SA;
+ *
+ *
+ */
+namespace oat\taoSync\model\testCenter;
+
+use oat\taoTestCenter\model\TestCenterService as BaseTestCenterService;
+use core_kernel_classes_Resource;
+use oat\taoSync\model\SyncService;
+use oat\taoSync\model\synchronizer\custom\byOrganisationId\testcenter\TestCenterByOrganisationId;
+use oat\oatbox\user\User;
+use oat\taoTestCenter\model\exception\TestCenterException;
+
+/**
+ * Class TestCenterService
+ * @package oat\taoSync\model\testCenter
+ * @author Aleh Hutnikau, <hutnikau@1pt.com>
+ */
+class TestCenterService extends BaseTestCenterService
+{
+
+    /**
+     * @param core_kernel_classes_Resource $testCenter
+     * @param User $user
+     * @param core_kernel_classes_Resource $role
+     * @return bool
+     * @throws \common_Exception
+     * @throws \core_kernel_persistence_Exception
+     * @throws \oat\taoTestCenter\model\exception\TestCenterException
+     */
+    public function assignUser(\core_kernel_classes_Resource $testCenter, User $user, \core_kernel_classes_Resource $role)
```

Comment:
This is what I mentioned in the previous PR, that it would be better to handle the roles in a separate constant, and don't hardcode them into your method so in that case it would be easier to extend the original method with new roles without modifying the logic itself (or duplicating code). What do you think about the following:
```
// taoTestCenter -> TestCenterService.php
    const ASSIGN_ROLE_MAP = [
        self::ROLE_TESTCENTER_ADMINISTRATOR => ProctorManagementService::PROPERTY_ADMINISTRATOR_URI,
        ProctorService::ROLE_PROCTOR => ProctorManagementService::PROPERTY_ASSIGNED_PROCTOR_URI,
    ];

    public function assignUser(\core_kernel_classes_Resource $testCenter, User $user, \core_kernel_classes_Resource $role)
    {
        // ...
        if (!isset(self::ASSIGN_ROLE_MAP[$role->getUri()])) {
            throw new TestCenterException(__('Role is not allowed.'));
        }
        
        $assignProperty = self::ASSIGN_ROLE_MAP[$role->getUri()];
        // ...
    }
```

```
// taoSync -> TestCenterService.php
class TestCEnterService extends BaseTestCenterService
{
    const ASSIGN_ROLE_MAP = [
        self::ROLE_TESTCENTER_ADMINISTRATOR => ProctorManagementService::PROPERTY_ADMINISTRATOR_URI,
        ProctorService::ROLE_PROCTOR => ProctorManagementService::PROPERTY_ASSIGNED_PROCTOR_URI,
        SyncService::TAO_SYNC_ROLE => SyncService::PROPERTY_ASSIGNED_SYNC_USER,
    ];

    public function assignUser($testCenter, $user, $role)
    {
        parent::assignUser($testCenter, $user, $role);

        $userResource = $this->getResource($user->getIdentifier());
        $organisationIdProperty = $this->getProperty(TestCenterByOrganisationId::ORGANISATION_ID_PROPERTY);
        $id = $this->getTestCenterOrganisationId($testCenter);

        $userResource->removePropertyValue($organisationIdProperty, $id);
        $userResource->setPropertyValue($organisationIdProperty, $id);
    }
}
```

So in that case no duplication, just the new logic after the original call. The `ASSIGN_ROLE_MAP` could be a class variable and do `array_merge()` instead of defining again.
It's just an idea, but I hope you see my point :)
Same could be applied to the `unassignUser()`.

Reply:
I think there is no any profit in hardcoding roles in a constant because logic of those services is different.